async-14: Prototyping Octree Visuals

pwinston opened the pull request:
# Description

* Wedge PR to get experimental octree-based image rendering in place.
* The experimental part is very WIP and will crash except for some specific test cases.
* For octree visuals set `NAPARI_ASYNC` to the path of a JSON config file:
    * If `octree_visuals` setting is missing or false: standard async.
    * If `octree_visuals` setting is true: octree visuals


## Type of change
- [x] Mainstream files changed (most guarded by `NAPARI_ASYNC`)
- [x] New experimental files and changes.

# Non-Experimental Changes

Safe/minimal change to `vispy_image_layer.py`:

The old `_on_data_change()` method took the data from `self.layer._data_view`, processed it, and passed the data to `self.node`. Now instead `_on_data_change()` calls a new method `_set_node_data(self, node, data):` which does the real work. The benefit is the derived class `VispyTiledImageLayer` can call `VispyImageLayer._set_node_data` over and over, to set the data for every node/tile.

|File|Comments|
|---|---|
|napari/_vispy/vispy_image_layer.py|Extract `_set_node_data()` from `_on_data_change()`|

# Non-Experimental Changes (Guarded)

These 5 changes are guarded by `NAPARI_ASYNC`:

|File|Comments|
|---|---|
|napari/_qt/qt_viewer.py|Create `dockerRender` = `QtRenderContainer`|
|napari/_qt/qt_main_window.py|Add `dockRender`|
|napari/_qt/layer_controls/qt_layer_controls_container.py|Use regular image controls for `OctreeImage`|
|napari/components/add_layers_mixin.py|Create `Image` or `OctreeImage`|
|napari/_vispy/utils.py|Map `OctreeImage` -> `VispyTiledImageLayer`|

# Experimental: Changes
|File|Comments|
|---|---|
|napari/components/experimental/chunk/_config.py|New Octree config setting|

# Experimental: New
|File|Comments|
|---|---|
|napari/_qt/experimental/qt_render.py|`QtRender` widget|
|napari/_qt/experimental/qt_render_container.py|`QtRenderContainer`|
|napari/_vispy/experimental/vispy_tiled_image_layer.py|Tiled rendering with multiple `ImageVisuals`|
|napari/layers/image/experimental/octree_image.py|Will replace `Image` class|
|napari/layers/image/experimental/_octree_image_slice.py| Will replace `ImageSlice` class|
|napari/layers/image/experimental/_octree.py|Octree Datastructure|

# Files

<img width="326" alt="pr_1707_files" src="https://user-images.githubusercontent.com/4163446/96532869-c405c080-125a-11eb-8c70-395e6500e69e.png">

# TiledImageVisual?

Above we are using `VispyTiledImageLayer` which creates one `ImageNode` per tile. On Vispy's Gitter David Hoese said that will not perform well. He said we should create a new single visual called `TiledImageVisual` which manages tiles with a texture atlas. So it can draw all the tiles in one draw command.

However, we are starting with `VispyTiledImageLayer` because it was simple. And we can upgrade it when we need the performance.

# References

* **Articles**
    * [Wikipedia: Octree](https://en.wikipedia.org/wiki/Octree)
    * [Loose Octrees](https://anteru.net/blog/2008/loose-octrees/)
    * [Introduction to Octree](https://www.gamedev.net/tutorials/programming/general-and-gameplay-programming/introduction-to-octrees-r3529/) (gamedev)
     * [Slippy Map](https://wiki.openstreetmap.org/wiki/Slippy_Map)
* **Python**
    * [ActiveState](https://code.activestate.com/recipes/498121-python-octree-implementation/)
    * [pyoctree](https://github.com/jcummings2/pyoctree)
* **C++**
    * [SimpleOctree](https://github.com/brandonpelfrey/SimpleOctree) (2013)
    * [PointCloudLibrary](https://github.com/PointCloudLibrary/pcl/tree/master/octree) (5000 lines, gold standard for correctness?)

# How has this been tested?
- [x] manual testing

pwinston committed:
Add vispy_chunked_image_layer

pwinston committed:
Create experimental VispyChunkedBaseLayer

we need to modify the image layer and base layer, so both need
to be in experimental

pwinston committed:
Introduce ChunkedImage layer...

pwinston committed:
Remove no longer need portion.

How that we have ChunkedImage layer.

pwinston committed:
Still need this stuff for non-chunked async

pwinston committed:
Test code for creating image pyramid

pwinston committed:
Extract function

pwinston committed:
Building the tree Part 1

pwinston committed:
Give OctreeNode row/col, for now at least

d-v-b commented on napari/layers/image/experimental/octree.py:
```diff
@@ -0,0 +1,197 @@
+    tiles = []
+
+    row = 0
+    while row < rows:
```
A lot of the tooling for dividing arrays into chunks exists in dask -- I'm pretty sure the thing you are doing here is done in n dimensions without while loops in the `slices_from_chunks` function in `dask.array.core` https://github.com/dask/dask/blob/45fd9092049a61c1d43067a0132574b3fe715153/dask/array/core.py#L206

pwinston replied:
Yeah. We haven't really had the discussion of using dask inside the renderer itself, as in everything turns into dask whether it started as dask or not. Ask of today I'm assuming we don't want to do that. But that's a valid conversation to have, even soon-ish.

d-v-b replied:
That function I referenced doesn't use dask arrays at all. It's just a convenience function to generate slices for dividing an array into chunks.

pwinston replied:
This specific code of building a tree from a single image is just for testing. Almost all multi-scale viewing will not go through this code, because it will be a pre-chunked format.

To fix #1300 though we might need to build a tree on the fly. If we did that I'd consider a better way to build the tree. Although the real solution there might be more about increment/amortized building. Because no matter how efficiently we did it, it might take too long if we did it in one gulp. At any rate not worrying about #1300 or on-the-fly yet, except just for this quick-n-dirty testing.

d-v-b replied:
That's all fine, I'm just letting you know that the _specific thing you are doing in this loop_ (slice an array into chunks) is already implemented via the dask core utils, so you could remove some of your code and just use / copy the dask function, if you want. For myself, whenever I'm breaking an array into little chunks I often find clean functions for that in the dask array core utils, even when I'm not working with dask arrays.

pwinston replied:
Yes that's good to know about, thanks.

pwinston committed:
Introduce ChunkedImageSlice and builds octree as a test

pwinston committed:
Big rename to reduce confusion

chunked = async,  uses ChunkLoader
octree = octree visuals

pwinston committed:
Rename config setting to 'octree_visuals'

pwinston committed:
Add block_timer, print less

pwinston committed:
Add OctreeImage.view_chunks and ChunkData class

pwinston committed:
Start VispyTiledImageLayer

This is the multi-ImageNode solution which will eventually be
placed by a VispyTiledImage texture-atlas (texture cache)
based solution.

pwinston committed:
Add VispyImageLayer2

These are proposed changes to VispyImageLayer2 which are
backward compatible. Such that we can just merge this
class in if all goes well.

pwinston committed:
Cleanup

pwinston committed:
Pass in self.node

pwinston committed:
Display just the root tile, placeholder for more

pwinston committed:
Renders but with fudged coordinates

pwinston committed:
Stub in QtRender GUI controls for octree

pwinston committed:
Debug tile borders and prelim set octree level

pwinston committed:
Octree level setting working via spinbox

pwinston committed:
Slightly better downsampling?

pwinston committed:
Comments and cleanup

pwinston committed:
Merge branch 'master' into feature/async-visual

pwinston committed:
Merge latest from image.py

pwinston committed:
Better comment

pwinston committed:
Undo these changes, async not using DebugMenu yet

pwinston committed:
Restore missing use_processes

pwinston committed:
Move QtRender to experimental

pwinston committed:
Should be no changes

pwinston committed:
Cleanup ASYNC feature toggle points

pwinston committed:
Make OctreeImage derive from Image

pwinston committed:
Change VispyImageLayer so it works with tiling

Works with VispyTiledImageLayer.

pwinston committed:
Derive OctreeImageSlice from ImageSlice

pwinston committed:
Need __init__ for manifest?

sofroniewn commented on napari/_qt/qt_viewer.py:
```diff
@@ -152,8 +153,12 @@ def __init__(self, viewer):
 
+        # Only created if using async rendering.
+        self.dockRender = self._create_render_dock_widget()
+
```
Can we make this off by default, something like

```suggestion
    if self.dockRender is not None:
        self.dockRender.setVisible(False)
```

Right now when I open my pathology example I just get an empty widget at the right. It's pretty easy to turn on with the shortcut

pwinston replied:
See latest commit, the render widget now only appears with octree visuals. It does nothing with non-octree visuals anyway so this is better.

sofroniewn replied:
nice, thanks!

sofroniewn commented on napari/_vispy/experimental/vispy_tiled_image_layer.py:
```diff
@@ -0,0 +1,105 @@
+    Future Work
+    -----------
+
+    This class will likely need to be replaced at some point. We plan to
+    write a TiledImageVisual class which stores the tile textures in an
+    atlas-like tile cache. Then in one draw command it can draw the all the
+    visible tiles in the right positions using texture coordinates.
+
+    One reason TiledImageVisual will be faster is today calling
+    ImageNode.set_data() causes a 15-20ms hiccup when that visual is next
+    drawn. It's not clear why the hiccup is so big, since transfering the
+    texture into VRAM should be under 2ms. However at least 4ms is spent
+    rebuilding the shader alone.
+
+    Since we might need to draw 10+ new tiles at one time, the total delay
+    could be up to 200ms total, which is too slow.
+
+    However, this initial approach with separate ImageVisuals should be a
+    good starting point, since it will let us iron out the octree
+    implementation and the tile placement math. Then we can upgrade
+    to the new TiledImageVisual when its ready, and everything should
+    still look the same, it will just be faster.
```
Good explanation, thanks!!

sofroniewn commented on napari/_vispy/vispy_image_layer.py:
```diff
@@ -66,14 +66,17 @@ def _data_astype(self, data, dtype):
             return
 
-        data = self.layer._data_view
+        self._set_node_data(self.node, self.layer._data_view)
+
+    def _set_node_data(self, node, data):
+        """Our self.layer._data_view has been updated, update our node.
+        """
```
I'm fine with this, is very clever for usage in tiled version

pwinston replied:
This works well since we are letting`VispyImageLayer` do all the pre-processing it was always doing. The derived `VispyTiledImageLayer` is just calling into `VispyImageLayer._set_node_data()` once per tile. So yeah this was nice.

If we do switch to a `TiledImageVisual` as David suggested, I'm not as positive who would do this preprocessing. Maybe `VispyImageLayer` can continue to do it, and it just sends the processed data to the `TiledImageVisual` rather than making an `ImageNode` out of it, like we do here. That would also be nice.

sofroniewn replied:
> If we do switch to a TiledImageVisual as David suggested, I'm not as positive who would do this preprocessing. Maybe VispyImageLayer can continue to do it, and it just sends the processed data to the TiledImageVisual rather than making an ImageNode out of it, like we do here. That would also be nice.

Indeed I was wondering this too, but we can address later

sofroniewn commented on napari/layers/image/experimental/octree.py:
```diff
@@ -0,0 +1,307 @@
+        else:
+            assert num_cols == len(row)
+
+    print(f"{num_rows} x {num_cols} = {num_rows * num_cols}")
```
is there a reason in this file you're using all the print statements instead of the logging ones? Doesn't need to be fixed now, but maybe something for the future

pwinston replied:
No real reason. For super temporary stuff I still sometimes use prints. But I migrate things to logging which I feel like we'll always want available. Always want to be able to toggle back on.

An example is like "I just loaded this data_id async" is probably something you always want to know, forever. But like while building the octree we might print a bunch of junk. But in the end maybe we just print the build time and the number of tiles or something minimal.

pwinston replied:
As we discussed once, there should be no prints in mainstream code. That could be CI enforced someday. Although every project I've worked on people just shout "your stuff is printing junk!" and someone gets rid of it. Has not been a huge issue. Yet stray prints pop up fairly regularly.

sofroniewn replied:
sounds good, we can address later. i like the idea of a CI test at some point

sofroniewn reviewed (approved):
Ok, this looks good to me. It's still heavily in-progress in certain areas deep in the experimental portion of the code, but it does make a very nice wedge into the main code base in a clean way. The experimental portion like the octree and tiled visual can now be refined in a much easier way.

As already noted in the long run there maybe opportunities to reuse some dask functionality within the octree code, but we can address that later too.

I've left a couple of minor comments, but would only like to see one thing addressed before merge, which is making the render dock widget not visible by default when using `NAPARI_ASYNC`. You could @pwinston add an async config variable that would make it on by default if you want, but otherwise it's just a shortcut key combo away!

pwinston committed:
Make QtRender only show up with octree

Change shortcute to Ctrl+Shift+R
